### PR TITLE
[MM-38274] Implement registerMessageWillBeUpdatedHook

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR adds a new `registerMessageWillBeUpdatedHook(newPost, oldPost)` to intercept edited messages before sending them to the server.

This was surprisingly missing given we offer similar functionality server-side and it was posing certain limitations: for example, this was brought up by a user that is currently implementing an e2e encryption plugin which needs a way to manipulate messages before they hit the server, whether they are new or updated.

#### Ticket

https://mattermost.atlassian.net/browse/MM-38274

#### Release Note

```release-note
Added new registerMessageWillBeUpdatedHook(newPost, oldPost) client-side plugin hook to intercept edited messages.
```

### CODE REVIEW

**Review by crspeller — APPROVED** [review]
Looks good!

**Review by hmhealey — APPROVED** [review]

**Comment by streamer45** [comment]
@furqanmlk 

I am thinking a smoke test around post editing with and without a plugin registering the implemented hook.
Feel free to reach out to me and I can provide a sample plugin to use during QA.

**Review by furqanmlk — APPROVED** [review]
Thanks @streamer45 for the help.
it's working as expected

Pasted Graphic

image

**Comment by streamer45** [comment]
@aaronrothschild Let me know if you have questions or if we can go ahead with this.

File: plugins/registry.js [hunk 1]
```diff
@@ -739,4 +739,31 @@ export default class PluginRegistry {
 
         return id;
     }
+
+    // Register a hook that will be called when a message is updated by the user before it
```
Comment (aaronrothschild) [inline plugins/registry.js:743]:
```suggestion
    // Register a hook that will be called when a message is edited by the user before it
```

**Review by aaronrothschild — APPROVED** [review]
LGTM

**Pull request merged**